src/treelearner/feature_histogram.hpp

```diff
@@ -211,30 +213,48 @@ class FeatureHistogram {
 
 private:
   /*!
-  * \brief Calculate the split gain based on sum_gradients and sum_hessians
+  * \brief Calculate the split gain based on regularized sum_gradients and sum_hessians
   * \param sum_gradients
   * \param sum_hessians
   * \return split gain
   */
   double GetLeafSplitGain(double sum_gradients, double sum_hessians) const {
-    return (sum_gradients * sum_gradients) / (sum_hessians);
+    if (sum_gradients < -lambda_l1_) {
+      double reg_sum_gradients = sum_gradients + lambda_l1_;
+      return (reg_sum_gradients * reg_sum_gradients) / (sum_hessians + lambda_l2_);
+    } else if (sum_gradients > lambda_l1_) {
+      double reg_sum_gradients = sum_gradients - lambda_l1_;
+      return (reg_sum_gradients * reg_sum_gradients) / (sum_hessians + lambda_l2_);
+    }
+    return 0.0f;
   }
 
   /*!
-  * \brief Calculate the output of a leaf based on sum_gradients and sum_hessians
+  * \brief Calculate the output of a leaf based on regularized sum_gradients and sum_hessians
   * \param sum_gradients
   * \param sum_hessians
   * \return leaf output
   */
   double CalculateSplittedLeafOutput(double sum_gradients, double sum_hessians) const {
-    return -(sum_gradients) / (sum_hessians);
+    if (sum_gradients < -lambda_l1_) {
+      return -(sum_gradients + lambda_l1_) / (sum_hessians + lambda_l2_);
+    } else if (sum_gradients > lambda_l1_) {
+      return -(sum_gradients - lambda_l1_) / (sum_hessians + lambda_l2_);
+    }
+    return 0.0f;
   }
```

Comment:
Looks like the assembly just uses `andps` to clear the sign bit, so that's pretty clean.